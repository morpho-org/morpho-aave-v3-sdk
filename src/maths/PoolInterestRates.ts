import { BigNumber, constants } from "ethers";

import { SECONDS_PER_YEAR } from "../constants/date";

import { MorphoAaveMath } from "./AaveV3.maths";

export interface PoolIndexesParams {
  /** The current pool supply index (in ray). */
  lastPoolSupplyIndex: BigNumber;

  /** The current pool borrow index (in ray). */
  lastPoolBorrowIndex: BigNumber;

  /** The current pool supply rate (in ray). */
  poolSupplyRatePerYear: BigNumber;

  /** The current pool borrow rate (in ray). */
  poolBorrowRatePerYear: BigNumber;

  /** The last update timestamp (in seconds). */
  lastUpdateTimestamp: BigNumber;

  /** The current timestamp (in seconds). */
  currentTimestamp: BigNumber;
}

export default class PoolInterestRates {
  __MATHS__ = new MorphoAaveMath();

  /**
   * Recompute the exact same logic as the Aave V3 protocol.
   * For supply index: https://github.com/aave/aave-v3-core/blob/9630ab77a8ec77b39432ce0a4ff4816384fd4cbf/contracts/protocol/libraries/logic/ReserveLogic.sol#L47
   * For borrow index: https://github.com/aave/aave-v3-core/blob/9630ab77a8ec77b39432ce0a4ff4816384fd4cbf/contracts/protocol/libraries/logic/ReserveLogic.sol#L73
   */
  public computePoolIndexes({
    lastPoolSupplyIndex,
    lastPoolBorrowIndex,
    poolSupplyRatePerYear,
    poolBorrowRatePerYear,
    lastUpdateTimestamp,
    currentTimestamp,
  }: PoolIndexesParams) {
    if (lastUpdateTimestamp.gte(currentTimestamp))
      return {
        newPoolSupplyIndex: lastPoolSupplyIndex,
        newPoolBorrowIndex: lastPoolBorrowIndex,
      };

    const newPoolSupplyIndex = this.__MATHS__.indexMul(
      lastPoolSupplyIndex,
      this._calculateLinearInterest(
        poolSupplyRatePerYear,
        lastUpdateTimestamp,
        currentTimestamp
      )
    );

    const newPoolBorrowIndex = this.__MATHS__.indexMul(
      lastPoolBorrowIndex,
      this._calculateCompoundedInterest(
        poolBorrowRatePerYear,
        lastUpdateTimestamp,
        currentTimestamp
      )
    );

    return {
      newPoolSupplyIndex,
      newPoolBorrowIndex,
    };
  }

  private _calculateLinearInterest(
    rate: BigNumber,
    lastUpdateTimestamp: BigNumber,
    currentTimestamp: BigNumber
  ) {
    const exp = currentTimestamp.sub(lastUpdateTimestamp);

    if (exp.isZero()) return this.__MATHS__.INDEX_ONE;

    return this.__MATHS__.INDEX_ONE.add(rate.mul(exp).div(SECONDS_PER_YEAR));
  }
  private _calculateCompoundedInterest(
    rate: BigNumber,
    lastUpdateTimestamp: BigNumber,
    currentTimestamp: BigNumber
  ) {
    const exp = currentTimestamp.sub(lastUpdateTimestamp);

    if (exp.isZero()) return this.__MATHS__.INDEX_ONE;

    const expMinusOne = exp.sub(1);
    const expMinusTwo = exp.gt(2) ? exp.sub(2) : constants.Zero;

    const basePowerTwo = this.__MATHS__
      .indexMul(rate, rate)
      .div(BigNumber.from(SECONDS_PER_YEAR).mul(SECONDS_PER_YEAR));

    const basePowerThree = this.__MATHS__
      .indexMul(basePowerTwo, rate)
      .div(SECONDS_PER_YEAR);

    const secondTerm = exp.mul(expMinusOne).mul(basePowerTwo).div(2);

    const thirdTerm = exp
      .mul(expMinusOne)
      .mul(expMinusTwo)
      .mul(basePowerThree)
      .div(6);

    return this.__MATHS__.INDEX_ONE.add(rate.mul(exp).div(SECONDS_PER_YEAR))
      .add(secondTerm)
      .add(thirdTerm);
  }
}
