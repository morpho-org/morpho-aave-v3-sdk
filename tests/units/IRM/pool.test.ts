import { BigNumber, constants } from "ethers";

import PoolInterestRates, {
  PoolIndexesParams,
} from "../../../src/maths/PoolInterestRates";

describe("PoolInterestRates", () => {
  let poolInterestRates: PoolInterestRates;
  let poolIndexes: PoolIndexesParams;
  beforeEach(() => {
    poolInterestRates = new PoolInterestRates();
    poolIndexes = {
      lastPoolSupplyIndex: BigNumber.from("1002790593020773538281770645"),
      poolSupplyRatePerYear: BigNumber.from("5308603204304240747807693"),
      lastPoolBorrowIndex: BigNumber.from("1005214768181497442705735506"),
      poolBorrowRatePerYear: BigNumber.from("17173304872362933916858043"),
      lastUpdateTimestamp: BigNumber.from("1680523835"),
      currentTimestamp: constants.Zero,
    };
  });
  describe("indexes", () => {
    it("REPL CASE 1: should increase correctly with time", () => {
      poolIndexes = {
        ...poolIndexes,
        currentTimestamp: poolIndexes.lastUpdateTimestamp.add(10),
      };
      expect(poolIndexes).toMatchSnapshot();

      // REPL CASE 1 solidity code result (check IRM/README for more info)
      const expectedNewPoolSupplyIndex = BigNumber.from(
        "1002790594708818108092852647"
      );
      const expectedNewPoolBorrowIndex = BigNumber.from(
        "1005214773655514349886614356"
      );

      const { newPoolSupplyIndex, newPoolBorrowIndex } =
        poolInterestRates.computePoolIndexes(poolIndexes);
      expect(newPoolSupplyIndex).toBnEq(expectedNewPoolSupplyIndex);
      expect(newPoolBorrowIndex).toBnEq(expectedNewPoolBorrowIndex);
    }); // REPL CASE 1

    it("REPL case 2: should not change if currentTimestamp == lastUpdateTimestamp", () => {
      poolIndexes = {
        ...poolIndexes,
        currentTimestamp: poolIndexes.lastUpdateTimestamp,
      };
      expect(poolIndexes).toMatchSnapshot();

      // REPL CASE 2 solidity code result (check IRM/README for more info)
      const expectedNewPoolSupplyIndex = BigNumber.from(
        "1002790593020773538281770645"
      );
      const expectedNewPoolBorrowIndex = BigNumber.from(
        "1005214768181497442705735506"
      );

      const { newPoolSupplyIndex, newPoolBorrowIndex } =
        poolInterestRates.computePoolIndexes(poolIndexes);
      expect(newPoolSupplyIndex).toBnEq(expectedNewPoolSupplyIndex);
      expect(newPoolBorrowIndex).toBnEq(expectedNewPoolBorrowIndex);

      // and the new pool indexes should be equal to the last indexes
      expect(newPoolSupplyIndex).toBnEq(poolIndexes.lastPoolSupplyIndex);
      expect(newPoolBorrowIndex).toBnEq(poolIndexes.lastPoolBorrowIndex);
    }); // REPL CASE 2
  });
});
