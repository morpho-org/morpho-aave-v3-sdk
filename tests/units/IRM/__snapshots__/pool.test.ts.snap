// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoolInterestRates indexes REPL CASE 1: should increase correctly with time 1`] = `
{
  "currentTimestamp": {
    "hex": "0x642ac245",
    "type": "BigNumber",
  },
  "lastPoolBorrowIndex": {
    "hex": "0x033f7e81c61aae35baddff52",
    "type": "BigNumber",
  },
  "lastPoolSupplyIndex": {
    "hex": "0x033d7d2af99d5e4834674e95",
    "type": "BigNumber",
  },
  "lastUpdateTimestamp": {
    "hex": "0x642ac23b",
    "type": "BigNumber",
  },
  "poolBorrowRatePerYear": {
    "hex": "0x0e3496b8fa414d5134f6bb",
    "type": "BigNumber",
  },
  "poolSupplyRatePerYear": {
    "hex": "0x046423f4d3417f25ed03cd",
    "type": "BigNumber",
  },
}
`;

exports[`PoolInterestRates indexes REPL case 2: should not change if currentTimestamp == lastUpdateTimestamp 1`] = `
{
  "currentTimestamp": {
    "hex": "0x642ac23b",
    "type": "BigNumber",
  },
  "lastPoolBorrowIndex": {
    "hex": "0x033f7e81c61aae35baddff52",
    "type": "BigNumber",
  },
  "lastPoolSupplyIndex": {
    "hex": "0x033d7d2af99d5e4834674e95",
    "type": "BigNumber",
  },
  "lastUpdateTimestamp": {
    "hex": "0x642ac23b",
    "type": "BigNumber",
  },
  "poolBorrowRatePerYear": {
    "hex": "0x0e3496b8fa414d5134f6bb",
    "type": "BigNumber",
  },
  "poolSupplyRatePerYear": {
    "hex": "0x046423f4d3417f25ed03cd",
    "type": "BigNumber",
  },
}
`;
